Make COR_PRF_DISABLE_OPTIMIZATIONS Allowable After Attach and non immutable

simonferquel opened the pull request:
Implement the changes discussed in #113921, making COR_PRF_DISABLE_OPTIMIZATIONS and COR_PRF_DISABLE_INLINING dynamically flippable, and applied consistently per module.

jkotas commented on src/coreclr/vm/vars.hpp:
```diff
@@ -548,27 +548,9 @@ inline bool CORDebuggerAttached()
-
-#ifdef DEBUGGING_SUPPORTED
-
-#define CORDisableJITOptimizations(dwDebuggerBits)        \
```
`CORDisableJITOptimizations` is used in number of places. Do those other places need to be update too now that it does not include the profiler flags?

Should we delete `CORDisableJITOptimizations` macro and replace its uses with `CORDebuggerAllowJITOpts`? Both macros are doing the same thing now.

simonferquel replied:
Something I am wondering, is if there is a real case where `DEBUGGING_SUPPORTED` is unset while `defined(PROFILING_SUPPORTED) || defined(PROFILING_SUPPORTED_DATA)` is true. Because in that case there is a change of behavior: we unconditionally return false while in the past, the profiler could disable JIT optimizations.

As the debug bits are always got from the module object, I wonder if it would make sense to remove the macro and have a member function on module instead - marked with inlining hints - that would look at debug flags + profiler de-optimization flag.
I actually made the profiler flag influence the debugBits initialization to satisfy that macro. It feels a bit wrong though as those bits are logically owned by the Debugger IIUC.

jkotas replied:
> As the debug bits are always got from the module object, I wonder if it would make sense to remove the macro and have a member function on module instead

Sounds good to me.

simonferquel replied:
While fixing that, I found a 2 instances of potentially wrong calls to the underlying "debugger-only" macro `CORDebuggerAllowJITOpts` - I think they should take into account profiler flag as well: 
- https://github.com/simonferquel/runtime/blob/d27ab2599124d9018704f5c1c8a22d75fbbea474/src/coreclr/debug/ee/debugger.inl?plain=1#L87C1-L87C49
- https://github.com/simonferquel/runtime/blob/d27ab2599124d9018704f5c1c8a22d75fbbea474/src/coreclr/debug/ee/debugger.inl?plain=1#L65C1-L65C64

Should I update those call sites as well ?

noahfalk replied:
> For the m_fHasOptimizedCode, it doesn't seem to be actually used anywhere, is it just for inspection to tell if the module is optimized? Could we add a note for its utility or remove it?

Agreed, that appears to be dead code. The field and its initialization can be deleted.
For the other one (DebuggerModule::HasAnyOptimizedCode()), yes I would update it.

simonferquel replied:
If the debugger macro should never be used outside of the method I introduced, shouldn't I move the macro code directly inside that method?

simonferquel replied:
Also, there is no "Profiler only" macro or method accessible outside of Module class, this is encapsulated in `AreJITOptimizationsDisabled` that checks both debugger and profiler bits.

noahfalk replied:
> shouldn't I move the macro code directly inside that method?

I like how the PR has it now with CORDisableJITOptimizations removed but CORDebuggerAllowJITOpts still existing. CORDebuggerAllowJITOpts also refers to g_CORDebuggerControlFlags and there are several other macros right next to it that show how other parts of g_CORDebuggerControlFlags are interpretted. I think that is useful.

@mdh1418 @jkotas - if you have a strong opinion otherwise do say so. I rarely have strong opinions on these subjective code refactoring choices but I want to give Simon clarity so this doesn't drag out.

simonferquel replied:
>  like how the PR has it now with CORDisableJITOptimizations removed but CORDebuggerAllowJITOpts still existing. CORDebuggerAllowJITOpts also refers to g_CORDebuggerControlFlags and there are several other macros right next to it that show how other parts of g_CORDebuggerControlFlags are interpretted. I think that is useful.

I actually removed it in my last batch of changes, but I can revert

jkotas replied:
If looks fine to me both with and without `CORDebuggerAllowJITOpts` macro. If you are going to delete `CORDebuggerAllowJITOpts` macro, you may want to also delete `CORDebuggerEnCMode` since that macro looks very one-off with `CORDebuggerAllowJITOpts` deleted.

simonferquel replied:
Ill revert it and add a comment about this only looking at debugger bits and that in most cases "AreJITOptimizationsDisabled" should be used

noahfalk reviewed (commented):
Looks good so far. I like your suggestion to refactor CORDisableJITOptimizations as a Module helper function so I expect that will change it a bit.

simonferquel commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -7865,6 +7865,13 @@ CorInfoInline CEEInfo::canInline (CORINFO_METHOD_HANDLE hCaller,
+    if (pOrigCallerModule->IsInliningDisabled())
+    {        
+        result = INLINE_NEVER;
+        szFailReason = "Inlining is disabled in the compiled method's module";
```
Oh actually yes, there is a reason:
Previously it was within 
```
if (CORProfilerPresent())
{
...
}
```
Now, the flag is set on the module on initialization and keeps consistent even if the profiler is detached - same behavior as the flag for JIT optimization.

simonferquel commented on src/coreclr/vm/ceeload.cpp:
```diff
@@ -504,6 +514,11 @@ void Module::SetDebuggerInfoBits(DebuggerAssemblyControlFlags newBits)
     m_dwTransientFlags |= (newBits << DEBUGGER_INFO_SHIFT_PRIV);
 
+    // if modules was initialized while Profiler disabled optimizations, disallow optimizations
+    if(m_dwTransientFlags & PROF_DISABLE_OPTIMIZATIONS){
+        m_dwTransientFlags &= ~DEBUGGER_ALLOW_JIT_OPTS_PRIV;
```
This change is removed

simonferquel commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -7865,6 +7865,13 @@ CorInfoInline CEEInfo::canInline (CORINFO_METHOD_HANDLE hCaller,
         goto exit;
     }
 
+    if (pOrigCallerModule->IsInliningDisabled())
```
Oh actually yes, there is a reason:
Previously it was within

if (CORProfilerPresent())
{
...
}
Now, the flag is set on the module on initialization and keeps consistent even if the profiler is detached - same behavior as the flag for JIT optimization.

simonferquel commented:
I could use some help diagnosing what is happening with https://dev.azure.com/dnceng-public/public/_build/results?buildId=1000924&view=logs&jobId=3e93a48f-6568-53f2-76b3-1285c183c39d&j=3e93a48f-6568-53f2-76b3-1285c183c39d&t=313dc435-bc2e-554b-c5b1-106790cdfaec.

tommcdon commented:
> I could use some help diagnosing what is happening with https://dev.azure.com/dnceng-public/public/_build/results?buildId=1000924&view=logs&jobId=3e93a48f-6568-53f2-76b3-1285c183c39d&j=3e93a48f-6568-53f2-76b3-1285c183c39d&t=313dc435-bc2e-554b-c5b1-106790cdfaec.

Adding @lewing as this looks like WASM testing

mdh1418 commented on src/coreclr/vm/ceeload.cpp:
```diff
@@ -478,6 +478,16 @@ void Module::Initialize(AllocMemTracker *pamTracker, LPCWSTR szName)
     m_dwTypeCount = 0;
     m_dwExportedTypeCount = 0;
     m_dwCustomAttributeCount = 0;
+#ifdef PROFILING_SUPPORTED
+    // set profiler related JIT flags
+    if(CORProfilerDisableInlining()){
+        m_dwTransientFlags |= PROF_DISABLE_INLINING;
+    }
+    if(CORProfilerDisableOptimizations()){
+        m_dwTransientFlags |= PROF_DISABLE_OPTIMIZATIONS;
+    }
+#endif //PROFILING_SUPPORTED
```
The below `#if defined(PROFILING_SUPPORTED) && !defined(DACCESS_COMPILE)` has a redundant check on `!defined(DACCESS_COMPILE)`. Can we remove the redundant check and group this under the same guard?

Nit to keep style consistent: space after if and `{` on next line

mdh1418 replied:
To clarify, is the intention that the bits corresponding to PROF_DISABLE_INLINING and PROF_DISABLE_OPTIMIZATIONS will only be set during initialization of the module?
Should we prevent them from being controlled by `SetTransientFlagInterlocked`?

simonferquel replied:
I think the debugger bits are in the same category. I'll check how / if it is done and apply the same logic.

simonferquel replied:
debugger flags are not protected either (I just checked). SetTransientFlagInterlocked being private, I think it is ok to leave it as it is in that PR and avoid adding more changes to it. Introducing a "DynamicFlagsMask" could be done in a followup.

noahfalk replied:
I think its fine that the checks are no stronger than what we do for debugging, ie no additional change needed for this.

mdh1418 commented on src/coreclr/vm/ceeload.h:
```diff
@@ -914,6 +916,31 @@ class Module : public ModuleBase
         return (m_dwTransientFlags & IS_EDIT_AND_CONTINUE) != 0;
     }
+    
+    BOOL IsInliningDisabledByProfiler() const {WRAPPER_NO_CONTRACT; SUPPORTS_DAC; return (m_dwTransientFlags & PROF_DISABLE_INLINING) != 0; }
```
nit
```suggestion
    BOOL IsInliningDisabledByProfiler() const { WRAPPER_NO_CONTRACT; SUPPORTS_DAC; return (m_dwTransientFlags & PROF_DISABLE_INLINING) != 0; }
```

mdh1418 commented on src/coreclr/vm/ceeload.h:
```diff
@@ -914,6 +916,31 @@ class Module : public ModuleBase
+        WRAPPER_NO_CONTRACT;
+        SUPPORTS_DAC;
+
+        #ifdef DEBUGGING_SUPPORTED
+        // check if debugger has disallowed JIT optimizations
+        if(!CORDebuggerAllowJITOpts(GetDebuggerInfoBits()))
+        {
+            return TRUE;
+        }
+        #endif
+
+        #if defined(PROFILING_SUPPORTED) || defined(PROFILING_SUPPORTED_DATA)
+        // check if profiler had disabled JIT optimizations when module was loaded
+        if(m_dwTransientFlags & PROF_DISABLE_OPTIMIZATIONS){
+            return TRUE;
+        }
+        #endif
```
To keep the style consistent, untab preprocessor directives + spacing after if + `{` next line.
```suggestion
#ifdef DEBUGGING_SUPPORTED
        // check if debugger has disallowed JIT optimizations
        if (!CORDebuggerAllowJITOpts(GetDebuggerInfoBits()))
        {
            return TRUE;
        }
#endif

#if defined(PROFILING_SUPPORTED) || defined(PROFILING_SUPPORTED_DATA)
        // check if profiler had disabled JIT optimizations when module was loaded
        if (m_dwTransientFlags & PROF_DISABLE_OPTIMIZATIONS)
        {
            return TRUE;
        }
#endif
```

mdh1418 replied:
If the check is under `defined(PROFILING_SUPPORTED) || defined(PROFILING_SUPPORTED_DATA)`, then should bit be set under the same condition? in ceeload.cpp its only `#ifdef PROFILING_SUPPORTED`

noahfalk replied:
> If the check is under defined(PROFILING_SUPPORTED) || defined(PROFILING_SUPPORTED_DATA), then should bit be set under the same condition? in ceeload.cpp its only #ifdef PROFILING_SUPPORTED

Using PROFILING_SUPPORTED only for the write and PROFILING_SUPPORTED || PROFILING_SUPPORTED_DATA for the read seems reasonable to me. PROFILING_SUPPORTED_DATA is used in the DAC where reading existing data structures is expected but initializing/writing them is not. The code would work correctly if PROFILING_SUPPORTED_DATA is added in the initialization path but it might create some confusion because it implies the init code might be executed by the DAC.

simonferquel replied:
I actually used the same condition as what was controlling the original macro. I am not familiar enough with the use of precompilation symbols in the runtime to have a strong opinion about it. If you think a change needs to be made, I'll be happy to do it.

noahfalk replied:
I don't think a change is needed :)

mdh1418 commented on src/coreclr/vm/jitinterface.cpp:
```diff
@@ -7865,6 +7865,13 @@ CorInfoInline CEEInfo::canInline (CORINFO_METHOD_HANDLE hCaller,
 
+    if (pOrigCallerModule->IsInliningDisabledByProfiler())
+    {        
+        result = INLINE_NEVER;
```
Why are we moving this outside of the PROFILING_SUPPORTED guard?
Also, why the change from INLINE_FAIL to INLINE_NEVER?

noahfalk replied:
From @simonferquel's past comment I'm guessing the intent was to move it outside the CORProfilerPresent() check because its possible to disable optimizations and then disconnect the profiler. I think the right spot would be inside the PROFILING_SUPPORTED build check but prior to the CORProfilerPresent().

mdh1418 commented on src/tests/profiler/native/dynamicjitoptimization/dynamicjitoptimization.cpp:
Could you please add spacing after `if` for consistency in readability?

mdh1418 commented on src/tests/profiler/dynamicoptimization/DynamicOptimization.cs:
Could you please keep the style consistent for readability? Please also see https://github.com/dotnet/runtime/blob/main/docs/coding-guidelines/coding-style.md

simonferquel replied:
I'll try to setup the linter correctly and run it on the PR changes.

jkotas commented on src/coreclr/debug/ee/debugger.cpp:
```diff
@@ -3379,12 +3379,10 @@ void Debugger::getVars(MethodDesc * md, ULONG32 *cVars, ICorDebugInfo::ILVarInfo
 
-    if (CORDisableJITOptimizations(bits))
+    if (md->GetModule()->AreJITOptimizationsDisabled())
 //    if (!CORDebuggerAllowJITOpts(bits))
```
```suggestion
```

simonferquel commented:
Per review comments, did:
- remove debug bits macro, encapsulating it in the helper method (actually enforcing both debugger and profiler bits are checked)
- moved IsInliningDisabledByProfiler check within PROFILING_SUPPORTED and returning the same INLINING_FAIL result as initially
- removed some redundant #ifdefs
- reviewed both C# and C++ formatting (it does not help that all clang-format / clang-tidy rules are removed at the root of the repo)

mdh1418 reviewed (commented):
Thank you for your effort! I think it looks good as long as we can confirm the test passes in CI

jkotas commented on src/coreclr/vm/ceeload.h:
```diff
@@ -915,6 +917,49 @@ class Module : public ModuleBase
+
+#ifdef DEBUGGING_SUPPORTED
+        // check if debugger has disallowed JIT optimizations
+        //
+        // original CORDebuggerAllowJITOpts macro definition (we encapsulate it here to force all JIT optimization checks to cover both debugger and profiler bits):
+        // #define CORDebuggerAllowJITOpts(dwDebuggerBits)           \
+        //     (((dwDebuggerBits) & DACF_ALLOW_JIT_OPTS)             \
+        //      ||                                                   \
+        //      ((g_CORDebuggerControlFlags & DBCF_ALLOW_JIT_OPT) && \
+        //       !((dwDebuggerBits) & DACF_USER_OVERRIDE)))
```
```suggestion
```
This comment is not useful. If somebody wants to dig into history, they can use source control.

simonferquel committed:
Make COR_PRF_DISABLE_OPTIMIZATIONS and COR_PRF_DISABLE_INLINING mutable flags

simonferquel committed:
On module initialization, capture the profiler JIT flags so they apply consistently on that module

simonferquel committed:
Change old macro usages and make them call Module::AreJITOptimizationsDisabled() instead

simonferquel committed:
Test for dynamic assignment of COR_PRF_DISABLE_OPTIMIZATIONS and COR_PRF_DISABLE_INLINING

simonferquel commented:
Last round of update:
- restored CORDebuggerAllowJITOpts macro, with a comment about preferring to call Module::AreJITOptimizationsDisabled()
- reorganized commit history (it was starting to be a mess)
- rebased to tentatively get any potential fixes for the seemingly unrelated build/test failures

simonferquel commented on src/coreclr/vm/ceeload.h:
```diff
@@ -915,6 +917,49 @@ class Module : public ModuleBase
         return (m_dwTransientFlags & IS_EDIT_AND_CONTINUE) != 0;
     }
 
+    BOOL IsInliningDisabledByProfiler() const
+    {
+        WRAPPER_NO_CONTRACT;
+        SUPPORTS_DAC;
+
+        return (m_dwTransientFlags & PROF_DISABLE_INLINING) != 0;
+    }
```
I think it is ok as is: if the profiler is not supported `m_dwTransientFlags` still exists and PROF_DISABLE_INLINING is never set. So the code is still returning the right thing

simonferquel commented on src/coreclr/vm/ceeload.cpp:
```diff
@@ -478,14 +478,23 @@ void Module::Initialize(AllocMemTracker *pamTracker, LPCWSTR szName)
+    }
 
-#if defined(PROFILING_SUPPORTED) && !defined(DACCESS_COMPILE)
+#if !defined(DACCESS_COMPILE)
```
As the guard was already there, I'd really prefer to do those cleanups as a follow-up if it is ok with you.

jkotas replied:
Well, this is very trivial cleanup - delete two obviously useless lines that you are changing anyway.

simonferquel commented:
Rebase brought new build failures - but it seems they are their on tip of main branch as well

noahfalk reviewed (approved):
LGTM, thanks @simonferquel!

simonferquel commented:
Should I worry about the build / test failures ? they all seem unrelated but could use confirmation

mdh1418 commented:
The Build Analysis pipeline should be green before we merge. It uses known issues to filter out failing checks, so the 3 other pipelines that are failing `Build Libraries Test Run checked coreclr linux x64 Release`, `Build osx-x64 Release AllSubsets Mono Minijit RuntimeTests minijit`, and `Build osx-x64 Release NativeAOT` have failures that either are
1) directly caused by this PR - Should fix
2) not related to this PR and there is no known issue linked to the failure - Should create a new issue to regex match the failure
3) not related to this PR and there is a known issue linked to the failure, but the regex matching couldn't match it - should either fix the regex matching or mention in the issue that the failure also occured in this PR

[Build Analysis Overview](https://github.com/dotnet/runtime/blob/a79768ed8b1a151d0daa002f92348ab4d47d73a0/docs/workflow/ci/failure-analysis.md)

mdh1418 reviewed (approved):
Looks good to me, thanks!

mdh1418 commented:
All the `Build Libraries Test Run checked coreclr linux x64 Release` are known issues https://github.com/dotnet/runtime/issues/114152 https://github.com/dotnet/runtime/issues/113883 and the `System.Net.Quic.Functional.Tests` is likely the same underlying problem (needs an issue).

jkotas commented on src/coreclr/vm/ceeload.cpp:
```diff
@@ -478,14 +478,23 @@ void Module::Initialize(AllocMemTracker *pamTracker, LPCWSTR szName)
     m_dwCustomAttributeCount = 0;
+#ifdef PROFILING_SUPPORTED
+    // set profiler related JIT flags
+    if (CORProfilerDisableInlining()){
+        m_dwTransientFlags |= PROF_DISABLE_INLINING;
+    }
+    if (CORProfilerDisableOptimizations()){
+        m_dwTransientFlags |= PROF_DISABLE_OPTIMIZATIONS;
+    }
```
```suggestion
    if (CORProfilerDisableInlining())
    {
        m_dwTransientFlags |= PROF_DISABLE_INLINING;
    }
    if (CORProfilerDisableOptimizations())
    {
        m_dwTransientFlags |= PROF_DISABLE_OPTIMIZATIONS;
    }
```
Follow the style used in the rest of the file

mdh1418 commented:
I think the OSX failures are some bad machine configuration.

@simonferquel, in addition to Jan's feedback, there are also conflicts in `src/tests/profiler/native/CMakeLists.txt` and `src/tests/profiler/native/classfactory.cpp` because 
https://github.com/dotnet/runtime/pull/114107 also added a profiler test

simonferquel committed:
Update src/coreclr/vm/ceeload.cpp

Co-authored-by: Jan Kotas <jkotas@microsoft.com>

simonferquel commented:
I rebased and fixed conflict. I also opened a dummy draft PR (removing an empty line of code) to compare build results (so that I do not introduce regression by feeling a flakky test issue)

simonferquel commented:
Build analysis is green, thanks for the help working on that one :)
